hal/armv7m/imxrt/timer.c

```diff
@@ -14,59 +14,99 @@
  */
 
 #include "../../../timer.h"
+#include "../../armv7m.h"
 #include "config.h"
 #include <arch/cpu.h>
 #include <arch/interrupts.h>
 
 
+enum { gpt_cr = 0, gpt_pr, gpt_sr, gpt_ir, gpt_ocr1, gpt_ocr2, gpt_ocr3, gpt_icr1, gpt_icr2, gpt_cnt };
+
+
 static struct {
 	intr_handler_t handler;
-	volatile time_t jiffies;
+	volatile u32 upper;
 	spinlock_t sp;
 
+	volatile u32 *base;
 	u32 interval;
 } timer_common;
 
 
-static int _timer_irqHandler(unsigned int n, cpu_context_t *ctx, void *arg)
+static time_t hal_timerCyc2Us(time_t ticks)
 {
-	(void)n;
-	(void)arg;
-	(void)ctx;
+	return (ticks * 1000 * 1000) / (GPT_FREQ / (GPT_PRESCALER * GPT_OSC_PRESCALER));
+}
```

Comment:
I may be wrong but,
running gcc with the -E option will yield such code:
```
static time_t hal_timerCyc2Us(time_t ticks)
{
 return (ticks * 1000 * 1000) / ((16 * 1000 * 1000) / ((16 / 8) * 8));
}
```
and simplifying
```
 return (ticks * 1000 * 1000) / (1000 * 1000) );
```
and simplifying it should be
```
return ticks;
```
i.e `ticks == us`
but the compiler performs this as first multiplying by 1000000 and then dividing also by 1000000.

The effect of this can be the occurrence of an overflow in the numerator if I count correctly after about 214 days.

Reply:
This statement should be optimized during compilation to `us = ticks`. It's here to allow a easy change to submicro precision if needed.

Should no optimization take place I think we're still ok, as both `ticks` and `us` are `time_t`, which is 64 bit type.

Reply:
Well that's exactly what the compiler does not optimize for `us = ticks`.
Code:
```
typedef unsigned long long time_t;

time_t hal_timerCyc2Us(time_t ticks)
{
	return (ticks * 1000 * 1000) / (1000 * 1000);
}
```
and disassembly:
```
Disassembly of section .text.hal_timerCyc2Us:

00000000 <hal_timerCyc2Us>:
typedef unsigned long long time_t;

time_t hal_timerCyc2Us(time_t ticks)
{
   0:   b5f8            push    {r3, r4, r5, r6, r7, lr}
        return (ticks * 1000 * 1000) / (1000 * 1000);
   2:   0145            lsls    r5, r0, #5
   4:   4a0e            ldr     r2, [pc, #56]   ; (40 <hal_timerCyc2Us+0x40>)
   6:   014c            lsls    r4, r1, #5
   8:   2300            movs    r3, #0
   a:   1a2d            subs    r5, r5, r0
   c:   ea44 64d0       orr.w   r4, r4, r0, lsr #27
  10:   ea4f 2645       mov.w   r6, r5, lsl #9
  14:   eb64 0401       sbc.w   r4, r4, r1
  18:   0264            lsls    r4, r4, #9
  1a:   ea44 54d5       orr.w   r4, r4, r5, lsr #23
  1e:   1835            adds    r5, r6, r0
  20:   eb41 0404       adc.w   r4, r1, r4
  24:   01ae            lsls    r6, r5, #6
  26:   01a7            lsls    r7, r4, #6
  28:   1b76            subs    r6, r6, r5
  2a:   ea47 6795       orr.w   r7, r7, r5, lsr #26
  2e:   eb67 0404       sbc.w   r4, r7, r4
  32:   1830            adds    r0, r6, r0
  34:   eb44 0101       adc.w   r1, r4, r1
  38:   f7ff fffe       bl      0 <__aeabi_uldivmod>    38: R_ARM_THM_CALL      __aeabi_uldivmod
}
  3c:   bdf8            pop     {r3, r4, r5, r6, r7, pc}
  3e:   bf00            nop
  40:   000f4240        .word   0x000f4240
```
and approximate calculations:
```
2^64 = 1,84467440737e+19

max_uul 2^64 ~ 1.8 * 10^19 - round down
max_uul 2^64 ~ 1.85 * 10^19 - round up
day_us ~ 86400 * 10^6 ~ 86 * 10^9
mul = 10^6
day_x_mul = day_us * mul = 86 * 10^9 * 10^6 = 86 * 10^15

max_days = max_uul / day_x_mul = (1.8 * 10^19 ) / (86 * 10^15) = (1.8 * 10^4) / 86 = 18000 / 86 ~ 209
max_days = max_uul / day_x_mul = (1.85 * 10^19) / (86 * 10^15) = (1.85 * 10^4) / 86 = 18500 / 86 ~ 215
```

Reply:
Ok, you convinced me, fixed, thanks. I've left 1000 factor to allow submicro precision (should allow for uptime of > 600 years).